### PR TITLE
[Interop] [SwiftToCxx] Using the hardcode noexcept flags on the compiler only to non throwing functions 

### PR DESCRIPTION
**Pull request by Robertorosmaninho**
This PR aims to modify the use of the hardcode flags that, so far, represented an unreal behavior from Swift functions that assume that every function doesn't throw an exception. The tests introduced by this [PR](https://github.com/apple/swift/compare/main...Robertorosmaninho:cxx-interop/SwiftToCxxErrorHandling?expand=1#59217) show the need for a new implementation to handle thrown functions.

This PR enable the tests [swift-functions-errors.swift](https://github.com/apple/swift/compare/main...Robertorosmaninho:interop/SwiftToCxxErrorHandling?expand=1#diff-95751b9663a6f895f190092c7a3a130983b4f5b21ee32c50dfbfe7f34a92a438) and [swift-functions-errors-execution.cpp](https://github.com/apple/swift/compare/main...Robertorosmaninho:interop/SwiftToCxxErrorHandling?expand=1#diff-6e56f9a7dcc273e66801494f1c061fea4dc4a6e90ace005f2c4968f652a51c09) and check that are correctly passing.

The main idea of this PR is to insert a condition to print the SWIFT_NOEXCEPT and the noexcept on the functions inside the bridging header files generated by Swift to interop with C++. This PR also introduces a trivial error type as an argument on a thrown function.

### CODE REVIEW

**Comment by egorzhdan** [comment]
@swift-ci please test

**Comment by Robertorosmaninho** [comment]
Hey @egorzhdan, I forgot about one test case, sorry about it! Can you please run the test-suit again?

**Comment by egorzhdan** [comment]
@swift-ci please test

File: include/swift/IRGen/IRABIDetailsProvider.h [hunk 1]
```diff
@@ -18,6 +18,7 @@
 #include "llvm/ADT/Optional.h"
 #include "llvm/ADT/STLExtras.h"
 #include "llvm/ADT/SmallVector.h"
+#include "swift/AST/Decl.h"
```
Comment (hyp) [inline include/swift/IRGen/IRABIDetailsProvider.h:21]: This include should be sorted, can you clang-format it?

File: lib/IRGen/IRABIDetailsProvider.cpp [hunk 2]
```diff
@@ -122,6 +125,32 @@ class IRABIDetailsProviderImpl {
     return {returnTy, {paramTy}};
   }
 
+  llvm::SmallVector<IRABIDetailsProvider::ABIAdditionalParam, 1>
+  getFunctionABIAdditionalParams(AbstractFunctionDecl *afd) {
+    llvm::SmallVector<IRABIDetailsProvider::ABIAdditionalParam, 1> params;
+    Mangle::ASTMangler mangler;
+
+    auto mangledName = mangler.mangleAnyDecl(afd, /*prefix=*/true);
+    for (auto &f : silMod->getFunctions()) {
```
Comment (hyp) [inline lib/IRGen/IRABIDetailsProvider.cpp:134]: Sorry, I know I suggested this approach of lowering to sil and finding the function initially, but it’s actually not the approach we should pursue. 
We just want to hook into SILGenModule, and it's expose its API we can then use from IRABIDetailsProvider.

For example:
// SomeHeader.h
SILFunction *getFunction(SILDeclRef ref);

// SomeFile.cpp (inside lib/SIL)
SILFunction *getFunction(SILDeclRef ref) {
  SILGenModule SGM(...);
  return SGM.getFunction(ref, NotForDef);
}

File: include/swift/IRGen/IRABIDetailsProvider.h (hunk 1 above)
Comment (Robertorosmaninho) [reply include/swift/IRGen/IRABIDetailsProvider.h]: Yes, sir!

File: lib/IRGen/IRABIDetailsProvider.cpp (hunk 2 above)
Comment (Robertorosmaninho) [reply lib/IRGen/IRABIDetailsProvider.cpp]: I used the `SILFunction.h` and `SILFunction.cpp` to place this function and it seems to work for the general case. However, this implementation sometimes crashes the compiler on the `function-availability.swift` test, but I couldn't reproduce the bug outside the lit test framework...

Comment (Robertorosmaninho) [reply lib/IRGen/IRABIDetailsProvider.cpp]: This is the output error: https://gist.github.com/Robertorosmaninho/276f9cf9888ce74e2e1a08ce41c3439d

File: lib/IRGen/IRABIDetailsProvider.cpp [hunk 3]
```diff
@@ -60,7 +64,7 @@ class IRABIDetailsProviderImpl {
 public:
   IRABIDetailsProviderImpl(ModuleDecl &mod, const IRGenOptions &opts)
       : typeConverter(mod),
-        silMod(SILModule::createEmptyModule(&mod, typeConverter, silOpts)),
+        silMod(SILModule::createEmptyModule(&mod, typeConverter, SILOptions())),
```
Comment (hyp) [inline lib/IRGen/IRABIDetailsProvider.cpp:67]: Please use the original `silOpts` here.

File: lib/IRGen/IRABIDetailsProvider.cpp [hunk 4]
```diff
@@ -122,6 +126,32 @@ class IRABIDetailsProviderImpl {
...
 
+  llvm::SmallVector<IRABIDetailsProvider::ABIAdditionalParam, 1>
+  getFunctionABIAdditionalParams(AbstractFunctionDecl *afd) {
+    llvm::SmallVector<IRABIDetailsProvider::ABIAdditionalParam, 1> params;
+    Mangle::ASTMangler mangler;
+
+    auto function = SILFunction::getFunction(SILDeclRef(afd), *silMod);
+    auto mangledName = mangler.mangleAnyDecl(afd, /*prefix=*/true);
+
+    if (function->getName() == mangledName) {
```
Comment (hyp) [inline lib/IRGen/IRABIDetailsProvider.cpp:137]: You don't need to compare the names here. You already know that the SIL function you have is this AST function :)

File: lib/PrintAsClang/DeclAndTypePrinter.cpp [hunk 5]
```diff
@@ -949,16 +968,35 @@ class DeclAndTypePrinter::Implementation
...
                                                 owningPrinter.typeMapping,
                                                 owningPrinter.interopContext);
+    llvm::SmallVector<DeclAndTypeClangFunctionPrinter::AdditionalParam, 2>
+        additionalParams;
+    auto ABIparams = owningPrinter.interopContext.getIrABIDetails()
+                         .getFunctionABIAdditionalParams(FD);
+    if (funcTy->isThrowing() && !ABIparams.empty()) {
+      if (ABIparams.pop_back_val().role ==
+          IRABIDetailsProvider::ABIAdditionalParam::ABIParameterRole::Self)
+        additionalParams.push_back(
```
Comment (hyp) [inline lib/PrintAsClang/DeclAndTypePrinter.cpp:978]: Can you share this code for constructing additional params with the section above where you create them too for C function?

File: lib/PrintAsClang/PrintClangFunction.cpp [hunk 6]
```diff
@@ -272,17 +272,23 @@ void DeclAndTypeClangFunctionPrinter::printFunctionSignature(
...
-      } else {
-        print(param.type, OptionalTypeKind::OTK_None, "_self",
-              /*isInOut*/ false, delegate);
+      if (param.role == AdditionalParam::Role::Self) {
+        CFunctionSignatureTypePrinterModifierDelegate delegate;
+        delegate.prefixIndirectParamValueTypeInC = [](raw_ostream &os) {
+          os << "SWIFT_CONTEXT ";
+        };
+        if (FD->hasThrows())
+          os << "SWIFT_CONTEXT ";
```
Comment (hyp) [inline lib/PrintAsClang/PrintClangFunction.cpp:281]: Do you need this here? We already add SWIFT_CONTEXT.

File: lib/IRGen/IRABIDetailsProvider.cpp [hunk 7]
```diff
@@ -122,6 +126,32 @@ class IRABIDetailsProviderImpl {
...
+      auto silFuncType = function->getLoweredFunctionType();
+      auto funcPointerKind =
+          FunctionPointerKind(FunctionPointerKind::BasicKind::Function);
+      auto signature =
+          Signature::getUncached(IGM, silFuncType, funcPointerKind);
+
+      for (unsigned i = 0; i < signature.getType()->getNumParams(); i++) {
+        auto set = signature.getAttributes().getAttributes(i);
+        if (set.hasAttribute(llvm::Attribute::AttrKind::SwiftSelf))
```
Comment (hyp) [inline lib/IRGen/IRABIDetailsProvider.cpp:146]: What about the SwiftErrorResult it looks like you're only getting one parameter here?

File: lib/PrintAsClang/PrintClangFunction.cpp (hunk 6 above)
Comment (Robertorosmaninho) [reply lib/PrintAsClang/PrintClangFunction.cpp:281]: At least on the swift-functions-errors* tests, the SWIFT_CONTEXT that we already have was never printed...

File: lib/IRGen/IRABIDetailsProvider.cpp (hunk 7 above)
Comment (Robertorosmaninho) [reply lib/IRGen/IRABIDetailsProvider.cpp:146]: Yeah, the signature at this level doesn't have a `llvm::Attribute::AttrKind::SwiftError`, so we can only assure that the `SwiftSelf` is there

File: lib/PrintAsClang/DeclAndTypePrinter.cpp (hunk 5 above)
Comment (Robertorosmaninho) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp]: I refactored it by pushing this piece of code to a separated function and replaced the code with a call to this function

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Review by hyp — APPROVED** [review]
LGTM

**Pull request merged**